Hunter Plugin: Fix salamander net display (Fixed)

nrodrigues0213 opened the pull request:
This is a bugfix for the problems described in #1169 and #1553.

**Problem**:  The hunter plugin stores the incorrect trap tile depending on the angle of the small tree. 

**Solution**: Whenever a trap change is detected that is not in our hash map (and thus null), we cannot assume it is our trap blindly. For each of these traps, we look up adjacent tiles in the hash map. If a tile is one of our keys, we have identified our trap. The changes also separate salamander handling to its own case body, in order to reduce the scope of these scans to salamanders only.

**Considerations**: While I could have shortened the code that was added, I wanted to ensure that there were no unnecessary calculations that could impact performance even in the smallest degree. That's why there are several if-statements and redundant code in each block. 

![after2](https://user-images.githubusercontent.com/25905465/38783826-df138de8-40d5-11e8-85cb-815d343f2dbf.png)

If the changes here could be improved in any way, let me know. I can add comments throughout the code if that would be useful.
**Note**: This was _only_ tested at green salamanders. If you have issues outside of green salamanders, leave a comment.

Kamielvf commented on runelite-client/src/main/java/net/runelite/client/plugins/hunter/HunterPlugin.java:
```diff
@@ -178,8 +195,44 @@ public void onGameObjectSpawned(GameObjectSpawned event)
+					// faces east. We have to look at adjacent tiles to locate the traps properly.
+					final WorldPoint objectTile = gameObject.getWorldLocation();
 
+					WorldPoint northTest = new WorldPoint(objectTile.getX(), objectTile.getY() + 1, objectTile.getPlane());
```
Wouldn't it be easier just to check what rotation the object has and determine it based on that?

nrodrigues0213 replied:
Good point. Working on an alternative solution using the rotation.

Adam- commented:
This seems unnecessarily complicated. Can you just translate the location based on object orientation before storing it in `traps` and then avoid this mess?

Joshua-F commented on runelite-client/src/main/java/net/runelite/client/plugins/hunter/HunterPlugin.java:
```diff
@@ -177,9 +199,101 @@ public void onGameObjectSpawned(GameObjectSpawned event)
+
+					switch (objectOrientation)
+					{
+						case 0:
```
Should store this case (and the ones below) as constants.

nrodrigues0213 commented:
Thank you for the comments. The last commit was entirely unfinished, my apologies. I've been looking at different approaches to fixing this using object orientation.

**Updated Problem**:
When a trap is placed, the tile of the net on the floor is stored.
When a trap is full, the entire game object shifts 0.5 game tiles in the direction of the net. The game object should occupy 2 tiles.
However, if a trap facing west or south changes state, the gameObject we are looking for can only be found on the tile of the small tree. This is why the algorithm cannot detect when the problematic trap shifts state to TRANSITION or FULL. 

The timer remains on the screen because onGameTick() only checks for occupation of a tile. The object occupies 2 tiles; yet, we can only detect changes in one for traps facing west or south. 

**Notes**:

- I have experimented with translating the trap tile when it is placed. However, this poses an issue: the timer remains on the screen even after the trap is dismantled. This occurs because we stored the tile of the small tree (for west and south facing traps) and onGameTick() doesn't check object ids, so it never removes dismantled traps from the hash map.

- An orientation check could be added for every new object that spawns, but that would result in many pointless function calls.

- An orientation check could be added to the TRANSITION and FULL cases, which would require much fewer lines.

- An ID check could be added to onGameTick().


If anything is unclear I can elaborate. I'll have much more time to work on this in a few hours. I'm more inclined to add orientation checks to the TRANSITION and FULL cases. I'm not sure about the performance costs of such ideas, so if you have suggestions please keep posting them. Thanks for your patience!

nrodrigues0213 commented:
I have rewritten the bug-fix using the orientation of the traps themselves.
In the process I have:

-  Added separate case for placing salamander traps. Changes the stored tile depending on the orientation of the hunter trap.
- Added new getTrapOrientation() function to keep the main function clean.
- Added new variable containsYoungTree to fix error with timers remaining on the screen.

If you have any further concerns, let me know.

Adam- commented:
Overall this looks ok. I think you could make a `getTrapLocation(GameObject)` to avoid some of the code duplication from this. Also maybe a general Angle abstraction and moving this game object orientation calculation out to a mixin perhaps would be best.

Adam- commented:
I pushed some unfinished work here for an Angle/Direction API. I think what you had before was reversed (or at least the comments were reversed). Update the plugin to work with the newer API.

nrodrigues0213 commented:
I've updated the hunter plugin to the new API and added getTrapLocation().  However, this is **not** functional yet. The client throws an exception when getOrientation() from the updated API is called:
```
Exception in thread "Thread-2" java.lang.AbstractMethodError: eb.getOrientation()Lnet/runelite/api/coords/Angle;
	at net.runelite.client.plugins.hunter.HunterPlugin.onGameObjectSpawned(HunterPlugin.java:154)
	at jdk.internal.reflect.GeneratedMethodAccessor87.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at com.google.common.eventbus.Subscriber.invokeSubscriberMethod(Subscriber.java:91)
	at com.google.common.eventbus.Subscriber$SynchronizedSubscriber.invokeSubscriberMethod(Subscriber.java:150)
	at com.google.common.eventbus.Subscriber$1.run(Subscriber.java:76)
	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:399)
	at com.google.common.eventbus.Subscriber.dispatchEvent(Subscriber.java:71)
	at com.google.common.eventbus.Dispatcher$PerThreadQueuedDispatcher.dispatch(Dispatcher.java:116)
	at com.google.common.eventbus.EventBus.post(EventBus.java:217)
	at ej.gameObjectsChanged(ej.java:248)
	at el.q(el.java:318)
	at el.y(el.java:261)
	at iy.j(iy.java:568)
	at fi.gi(fi.java:7375)
	at client.ft(client.java:3121)
	at client.at(client.java:1580)
	at br.g(br.java:382)
	at br.copy$run(br.java:361)
	at br.run(br.java:60)
	at java.base/java.lang.Thread.run(Thread.java:844)
```
I don't have enough experience to know where to move from here. What do I need to learn to handle these issues in the future so I could be more helpful?

Kamielvf commented on runelite-client/src/main/java/net/runelite/client/plugins/hunter/HunterPlugin.java:
```diff
@@ -132,43 +134,42 @@ public void onGameObjectSpawned(GameObjectSpawned event)
-					WorldPoint trapLocation = gameObject.getWorldLocation();
+					WorldPoint translatedTrapLocation = trapLocation;
 
 					switch (trapOrientation)
```
I would atleast support all orientations here, even if they are not used ingame.

nrodrigues0213 replied:
Good idea. I'll add the other orientations once I can test this again.

nrodrigues0213 commented:
mvn clean install fixed the problem I was having. Translating the other 2 directions brings back the original issue we were having, with timers not transitioning properly. I can add the 2 directions with a translation of 0 tiles as a formality, but I think that is unnecessary. I've flipped the directions; as Adam mentioned earlier, I had the directions flipped. Tested this with all 4 directions and it works!

Adam- commented:
So, to be clear, both south and west require no translations?

nrodrigues0213 commented:
That is correct. 

If you add translations to south and west, the timers break and never turn green. 

- North, East: Translated trap location
- South, West: Original trap location
